packages/core/echo/echo-pipeline/src/automerge/collection-synchronizer.ts

```diff
@@ -0,0 +1,197 @@
+//
+// Copyright 2024 DXOS.org
+//
+
+import { Event, scheduleTask, scheduleTaskInterval } from '@dxos/async';
+import { next as am } from '@dxos/automerge/automerge';
+import type { DocumentId, PeerId } from '@dxos/automerge/automerge-repo';
+import { Resource, type Context } from '@dxos/context';
+import { defaultMap } from '@dxos/util';
+
+const MIN_QUERY_INTERVAL = 5_000;
+
+const POLL_INTERVAL = 30_000;
+
+export type CollectionSynchronizerParams = {
+  sendCollectionState: (collectionId: string, peerId: PeerId, state: CollectionState) => void;
+  queryCollectionState: (collectionId: string, peerId: PeerId) => void;
+  shouldSyncCollection: (collectionId: string, peerId: PeerId) => Promise<boolean>;
+};
+
+/**
+ * Implements collection sync protocol.
+ */
+export class CollectionSynchronizer extends Resource {
+  private readonly _sendCollectionState: CollectionSynchronizerParams['sendCollectionState'];
+  private readonly _queryCollectionState: CollectionSynchronizerParams['queryCollectionState'];
+  private readonly _shouldSyncCollection: CollectionSynchronizerParams['shouldSyncCollection'];
+
+  /**
+   * CollectionId -> State.
+   */
+  private readonly _perCollectionStates = new Map<string, PerCollectionState>();
+
+  private readonly _connectedPeers = new Set<PeerId>();
+
+  public readonly remoteStateUpdated = new Event<{ collectionId: string; peerId: PeerId }>();
+
+  constructor(params: CollectionSynchronizerParams) {
+    super();
+    this._sendCollectionState = params.sendCollectionState;
+    this._queryCollectionState = params.queryCollectionState;
+    this._shouldSyncCollection = params.shouldSyncCollection;
+  }
+
+  protected override async _open(ctx: Context): Promise<void> {
+    scheduleTaskInterval(
+      this._ctx,
+      async () => {
+        for (const collectionId of this._perCollectionStates.keys()) {
+          this.refreshCollection(collectionId);
+        }
+      },
+      POLL_INTERVAL,
+    );
+  }
+
+  getRegisteredCollectionIds(): string[] {
+    return [...this._perCollectionStates.keys()];
+  }
+
+  getLocalCollectionState(collectionId: string): CollectionState | undefined {
+    return this._getPerCollectionState(collectionId).localState;
+  }
+
+  setLocalCollectionState(collectionId: string, state: CollectionState) {
+    this._getPerCollectionState(collectionId).localState = state;
+
+    queueMicrotask(async () => {
+      for (const peerId of this._connectedPeers) {
+        await this._refreshInterestedPeers(collectionId, peerId);
```

Comment:
should it be async?

Reply:
not sure, but I generally prefer something to be async, unless there's a strong need for it to be synchronous (e.g. UI/signal-related code)

Reply:
I think the opposite unless it's smth very generic like `open/close` on `resource/extension`.
`async` is always more suspicious (can it fail, how long can it take, can it hang, can it timeout?) and makes the client code harder like [here with Promise.all](https://github.com/dxos/dxos/pull/7244/files#diff-755dd7515a12273099fb49476cbda4caaf738263c1f702b0138ee35897994dbeR166)

it also looks funny that send and query in `CollectionSynchronizerParams` are sync, but this one is async. you'd usually expect the first two to be async